std/math/exponential.d

```diff
@@ -826,37 +826,211 @@ if (isUnsigned!F && isUnsigned!G && isUnsigned!H)
+            if (c <= 0x10000)
+            {
+                T result=a*b;
+                return result%c;
             }
+            version (D_InlineAsm_X86_64)
+            {
+                uint low=void;
+                uint high=void;
+
+                asm pure @trusted nothrow @nogc
+                {
+                    mov EAX,a;
+                    mul b;
+                    mov low,EAX;
+                    mov high,EDX;
+                }
 
-            T result = 0, tmp;
+                if (high >= c)
+                {
+                    high%=c;
+                }
+
+                if (high == 0)
+                {
+                    return low%c;
+                }
 
-            b %= c;
-            while (a > 0)
+                asm pure @trusted nothrow @nogc
+                {
+                    mov EAX,high;
+                    xor EDX,EDX;
+                    div c;
+                    mov high,EDX;
+
+                    mov EAX,low;
+                    mov EDX,high;
+                    div c;
+                    mov low,EDX;
+                }
+
+                return low;
+            }
+            else version (LDC)
             {
-                if (a & 1)
-                    result = addmod(result, b, c);
+                uint low,high;
 
-                a >>= 1;
-                b = addmod(b, b, c);
+                asm pure @trusted nothrow @nogc
+                {
+                    "mull %[b]"
+                    :"=a"(low),"=d"(high)
+                    :[a]"a"(a),[b]"r"(b)
+                    :"cc";
+                }
+
+                if (high >= c)
+                {
+                    high%=c;
+                }
+
+                if (high == 0)
+                {
+                    return low%c;
+                }
+
+                asm pure @trusted nothrow @nogc
+                {
+                    "divl %[c]"
+                    :"=a"(low),"=d"(high)
+                    :[a]"a"(low),"d"(high),[c]"r"(c)
+                    :"cc";
+                }
+
+                return low;
             }
+            else{
+                static T addmod (T x,T y,T m)
+                {
+                    y=m-y;
+                    if (x >= y)
+                        return x-y;
+                    else
+                        return m-y+x;
+                }
+                T result=0;
+                b%=c;
+                while (a > 0)
+                {
+                    if (a&1)
+                        result=addmod(result,b,c);
+                    a>>=1;
+                    b=addmod(b,b,c);
+                }
 
-            return result;
+                return result;
+            }
         }
         else
         {
+            alias DoubleT=AliasSeq!(void,ushort,uint,void,ulong)[T.sizeof];
             DoubleT result = cast(DoubleT) (cast(DoubleT) a * cast(DoubleT) b);
-            return result % c;
+            return cast(T)(result%c);
```

Comment:
Redundant cast here. VRP ensures that the type is implicitly cast to `typeof(c)`.